lib/cachedir.js

```diff
@@ -99,22 +120,23 @@ Cachedir.prototype.download = function(url, force, callback) {
               log.verbose('locked ' + output_file + ', starting downloading...');
 
               // get the data
-              r = request(url),
+              var r = request(url);
 
-                // do not continue on download errors
-                r.on('error', function(err) {
-                  return fs.close(fd, function() {
-                    fs.unlink(output_file);
-                    callback(err);
-                  });
+              // do not continue on download errors
+              r.on('error', function(err) {
+                return fs.close(fd, function() {
+                  cleanup();
```

Comment:
I feel like there could be a race condition here, where an error on request download results in the writablestream's fd being closed, and asynchronously after that this handler then invokes callback with error. But if the fd close were to trigger the writer.on('close', ...) handler below before the callback is invoked here, that handler could end up reporting success (no err and a path to the broken cache file). I haven't been able to trigger that locally though.